### PR TITLE
Feature extractor (NetHackNet) integration

### PR DESCRIPTION
**Pull request by Minchan-Kim**
I borrowed NetHackNet as a feature extractor and tested that it works with pop-art algorithm (actually it is more like IMPALA now). Currently it only woks on a CPU.

### CODE REVIEW

**Comment by yycho0108** [comment]
Current branch has conflicts with `main`. Please rebase from `origin/main`

**Review by yycho0108 — CHANGES_REQUESTED** [review]
Current branch has conflicts with `main`. Please rebase from `origin/main`. Otherwise, it's really hard to see which parts you've changed.

You'll need commits from 17ed05f8f33596e4a6afbbe5ae70b3e13cb8de21 to 3688f923ea5ffe6fa7faf300290c9fa8fd373314.
If you don't know how to do this, **backup your branch**, and then look up how to do `git rebase --onto`.

**Comment by Minchan-Kim** [comment]
Does it look good now?

File: feature_extractor.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        glyphs = torch.flatten(glyphs, 0, 1)  # Merge time and batch.
+
+        # -- [B' x F]
+        blstats = blstats.view(T * B, -1).float()
+
+        # -- [B x H x W]
+        glyphs = glyphs.long()
+        # -- [B x 2] x,y coordinates
+        coordinates = blstats[:, :2]
+        # TODO ???
```
Comment (yycho0108) [inline feature_extractor.py:202]:
```suggestion
```

File: feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        # coordinates[:, 0].add_(-1)
```
Comment (yycho0108) [inline feature_extractor.py:203]:
```suggestion
```

File: feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        blstats = blstats.view(T * B, -1).float()
+        # -- [B x K]
+        blstats_emb = self.embed_blstats(blstats)
+
+        assert blstats_emb.shape[0] == T * B
+
+        reps = [blstats_emb]
+
+        # -- [B x H' x W']
+        crop = self.crop(glyphs, coordinates)
```
Comment (yycho0108) [inline feature_extractor.py:215]: Why are we cropping?

File: feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+        # CNN crop model.
+        # -- [B x K x W' x H']
+        crop_emb = crop_emb.transpose(1, 3)  # -- TODO: slow?
+        # -- [B x W' x H' x K]
+        crop_rep = self.extract_crop_representation(crop_emb)
+
+        # -- [B x K']
+        crop_rep = crop_rep.view(T * B, -1)
+        assert crop_rep.shape[0] == T * B
```
Comment (yycho0108) [inline feature_extractor.py:231]: No need for obvious asserts
```suggestion
```

File: feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+    def forward(self, env_outputs, core_state, done):
+        # -- [T x B x H x W]
+        glyphs = env_outputs["glyphs"]
+
+        # -- [T x B x F]
+        blstats = env_outputs["blstats"]
+
+        T, B, *_ = glyphs.shape
+
```
Comment (yycho0108) [inline feature_extractor.py:191]: Please don't make extraneous newlines. It makes the code unnecessarily long.
My recommendation is to make newlines ONLY when there's a logical break between the code.
For instance:
```python
# inference
a = process_image()
b = extract_features(a)
c = compute_loss(b, y)
# -- [[ newline here but not elsewhere ]] --
# optimization
c.backward()

File: feature_extractor.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        reps = [blstats_emb]
+
+        # -- [B x H' x W']
+        crop = self.crop(glyphs, coordinates)
+
+        # print("crop", crop)
+        # print("at_xy", glyphs[:, coordinates[:, 1].long(), coordinates[:, 0].long()])
+
+        # -- [B x H' x W' x K]
+        crop_emb = self._select(self.embed, crop)
```
Comment (yycho0108) [inline feature_extractor.py:221]: Select what? The name should be more informative.
```suggestion
        crop_emb = self._select(self.embed, crop)
```

File: feature_extractor.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                out_channels=out_channels[i],
+                kernel_size=(F, F),
+                stride=S,
+                padding=P,
+            )
+            for i in range(L)
+        ]
+
+        self.extract_crop_representation = nn.Sequential(
+            *interleave(conv_extract_crop, [nn.ELU()] * len(conv_extract))
```
Comment (yycho0108) [inline feature_extractor.py:139]: My general preference is to make a _block_, instead of doing something like this, which feels like an awkward workaround.
`class Conv2dWithActivation([...], activation_fn='...')`
But this is just my preference.

File: feature_extractor.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+            [
+                self.width_grid[None, :, :] + x_shift[:, None, None],
+                self.height_grid[None, :, :] + y_shift[:, None, None],
+            ],
+            dim=3,
+        )
+
+        # TODO: only cast to int if original tensor was int
+        return (
+            torch.round(F.grid_sample(inputs, grid, align_corners=True))
```
Comment (yycho0108) [inline feature_extractor.py:61]: it would be nice to know what this is doing. Why do you need `grid_sample` as opposed to
`torch.narrow()` if all you're trying to express is `y = x[i0:i1, j0:j1]...`? What are you cropping and how?

File: feature_extractor.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        # glyphs_emb = self.embed(glyphs)
+        # -- [B x K x W x H]
+        glyphs_emb = glyphs_emb.transpose(1, 3)  # -- TODO: slow?
+        # -- [B x W x H x K]
+        glyphs_rep = self.extract_representation(glyphs_emb)
+
+        # -- [B x K']
+        glyphs_rep = glyphs_rep.view(T * B, -1)
+
+        assert glyphs_rep.shape[0] == T * B
```
Comment (yycho0108) [inline feature_extractor.py:246]:
```suggestion
```

File: feature_extractor.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+        # -- [B x K'']
+        reps.append(glyphs_rep)
+
+        st = torch.cat(reps, dim=1)
```
Comment (yycho0108) [inline feature_extractor.py:251]: st= abbreviation for ?

File: feature_extractor.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+                core_state = tuple(nd * s for s in core_state)
+                output, core_state = self.core(input.unsqueeze(0), core_state)
+                core_output_list.append(output)
+            core_output = torch.flatten(torch.cat(core_output_list), 0, 1)
+        else:
+            core_output = st
+
+        return core_output, core_state
+
+        # -- [B x A]
```
Comment (yycho0108) [inline feature_extractor.py:274]: Remove unused commented block

File: feature_extractor.py [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+            nn.Linear(out_dim, self.h_dim),
+            nn.ReLU(),
+            nn.Linear(self.h_dim, self.h_dim),
+            nn.ReLU(),
+        )
+
+        if self.use_lstm:
+            self.core = nn.LSTM(self.h_dim, self.h_dim, num_layers=1)
+
+        self.policy = nn.Linear(self.h_dim, self.num_actions)
```
Comment (yycho0108) [inline feature_extractor.py:166]: Do you need / use `self.policy` and `self.baseline`??

**Review by yycho0108 — APPROVED** [review]
Seems generally ok, left mostly stylistic comments.
* No need to address all comments if you do not have time.
* Please merge the PR _yourself_ after addressing the comments that you'd like to address.

File: feature_extractor.py (hunk 1 above)
Comment (Minchan-Kim) [reply feature_extractor.py:202]: -> FB

File: feature_extractor.py (hunk 2 above)
Comment (Minchan-Kim) [reply feature_extractor.py:203]: -> FB

File: feature_extractor.py (hunk 4 above)
Comment (Minchan-Kim) [reply feature_extractor.py:231]: -> FB, but I erased that line anyway.

File: feature_extractor.py (hunk 5 above)
Comment (Minchan-Kim) [reply feature_extractor.py:191]: -> FB

File: feature_extractor.py (hunk 6 above)
Comment (Minchan-Kim) [reply feature_extractor.py:221]: -> FB

File: feature_extractor.py (hunk 7 above)
Comment (Minchan-Kim) [reply feature_extractor.py:139]: -> FB

File: feature_extractor.py (hunk 9 above)
Comment (Minchan-Kim) [reply feature_extractor.py:246]: -> FB, but I erased that line anyway.

File: feature_extractor.py (hunk 10 above)
Comment (Minchan-Kim) [reply feature_extractor.py:251]: state, I guess.

File: feature_extractor.py (hunk 11 above)
Comment (Minchan-Kim) [reply feature_extractor.py:274]: -> FB

File: feature_extractor.py (hunk 12 above)
Comment (Minchan-Kim) [reply feature_extractor.py:166]: No.

**Comment by Minchan-Kim** [comment]
About most of your comments (regarding the coding style), I'm ignoring them since I did not write this code. "FB" in my comment means that you should ask Facebook Research, not me... Did you read my previous comment saying "I BORROWED NetHackNet ~"? Maybe it was not clear since I didn't specify where exactly I borrowed the code from. My purpose was to take a working network so that we can test pop-art or A2C quickly.

File: feature_extractor.py (hunk 3 above)
Comment (Minchan-Kim) [reply feature_extractor.py:215]: Read my answer to your question below about F.grid_sample.

File: feature_extractor.py (hunk 8 above)
Comment (Minchan-Kim) [reply feature_extractor.py:61]: Below is the explanation I found in the "The NetHack Learning Environment" (NeurIPS 2020) paper. 
"We apply a ConvNet (red) to all visible glyph embeddings as well as another ConvNet (blue) to the 9×9 crop of glyphs around the agent to create a dedicated egocentric representation for improved generalization."
I spent a lot of time to understand what are they doing by grid_sample, but it does not make sense to me also... They are upsampling 9 by 9 grid to 21 by 79 grid, but why? And what's the meaning of interpolation regarding the fact that the numbers are glyphs, not RGB values (non-integer values will lose their meaning)?

**Comment by yycho0108** [comment]
@Minchan-Kim I wasn't sure how much of the code was directly copy-pasted (vs. re-implemented). I guess that makes sense. Bad code is still bade code, but I suppose you're not to blame. :shrug:
